### ISSUE TITLE
Passing strings AssemblyScript <> Native

### ISSUE THREAD

**Issue by alvarowolfx** [wasm3/wasm3-arduino#1]
I'm new to WebAssembly but really excited to build some demos in the IoT context, so pardon if I'm missing something obvious. I tried the basic examples with an ESP32, PlatformIO, and the wasm3 repo and now I'm trying to build a bridge to more functions on the Arduino SDK. 

I'm trying to pass strings to the C part of code, but no matter what I try, only the first letter of the string arrives on the function. The second param with the length is correct but I couldn't get the full string.

I'm using AssemblyScript as my WebAssembly high-level API. I tried with TinyGo too, but I got compilation errors when using strings types.

I add this to `m3_arduino_api` as an example to bridge the Serial.println
```cpp
m3ApiRawFunction(m3_arduino_log)
{
  m3ApiGetArgMem(const char *, out)
  m3ApiGetArg(uint32_t, out_len)

  char output[out_len + 1];
  memcpy(output, out, out_len);
  Serial.println(output);

  m3ApiSuccess();
}

/* added to m3_LinkArduino function */ 
m3_LinkRawFunction(module, arduino, "log", "v(*i)", &m3_arduino_log);
```

My Assemblyscript code
```ts
@external("arduino", "millis")
declare function millis(): u32;

@external("arduino", "delay")
declare function delay(ms: u32): void;

@external("arduino", "pinMode")
declare function pinMode(pin: u32, mode: u32): void;

@external("arduino", "digitalWrite")
declare function digitalWrite(pin: u32, value: u32): void;

@external("arduino", "getPinLED")
declare function getPinLED(): u32;

@external("arduino", "log")
declare function println(output: string, len: u32): void;

function log(out: string): void {
  println(out, out.length)
}

const LOW: u32 = 0;
const HIGH: u32 = 1;

const INPUT: u32 = 0x0;
const OUTPUT: u32 = 0x1;
const INPUT_PULLUP: u32 = 0x2;

let LED: u32 = -1;

function setup(): void {
  LED = getPinLED();
  pinMode(LED, OUTPUT);
}

function run(): void {
  log('From Web Assembly')
  log('Hello\0')
  digitalWrite(LED, HIGH);
  delay(1000);
  digitalWrite(LED, LOW);
  delay(1000);
}

/*
 * Entry point
 */
export function _start(): void {
  setup();
  while (1) run();
}
```

**Comment by MaxGraey** [comment]
WebAssembly in current stage can pass only numbers between wasm module and host so when you write:
```ts
declare function println(output: string, len: u32): void;
```
it just the same as
```ts
declare function println(output: usize, len: u32): void;
```
where `output` is pointer (offset) to linear memory.
So you should read it on host. On web AssemblyScript and other langs provide some glue code (bindgen tools). You could see how read (decode) strings [here](https://github.com/AssemblyScript/assemblyscript/blob/master/lib/loader/index.js#L43) for example. AS use UTF-16LE, Rust and TinyGo - UTF-8 and required different approaches. Also I recommend investigate [memory layouts](https://docs.assemblyscript.org/details/memory#internals) which used in AS.

@vshymanskyy I guess need more advanced examples with passing arrays & strings.

**Comment by vshymanskyy** [comment]
@MaxGraey We already have API to access memory from Raw Functions.
But UTF-16LE support would be out-of-scope here. And I believe, `UTF-16LE` is the problem that @alvarowolfx is facing.

**Comment by alvarowolfx** [comment]
Gotcha. I also tried using `UInt8Array` and `String.UTF8.encode` from Assemblyscript, but the ESP32 always crash with invalid memory access just by having those referenced on code, which as I mentioned before, I have no clue why because I new to this 😅 

I'm trying right now to use linear memory raw access, but when I compile Assemblyscript with `--importMemory` flag, Wasm3 fails to start  with error `Fatal: m3_LoadModule: unallocated linear memory`

**Comment by MaxGraey** [comment]
Also, you could transform UTF16-LE to UTF8/Latin1:
```ts
declare function println(ptr: usize, len: i32): void;

function log(str: string): void {
  println(String.UTF8.encode(str, true), String.UTF8.byteLength(str, true)); // true if need null temination
}
```

**Comment by vshymanskyy** [comment]
Wouln't that include all of the conversion code into the wasm file?

**Comment by MaxGraey** [comment]
`--importMemory flag` mean you should setup memory via host, not sure wasm3 support this. But why you need this? AS by default export memory.

**Comment by alvarowolfx** [comment]
I tried your two proposed changes @MaxGraey, but as I mentioned before when I use String.UTF8.encode method, the ESP32 crashes. Here is an example output: 

```
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.
Core 0 register dump:
PC      : 0x4008124f  PS      : 0x00060630  A0      : 0x80081362  A1      : 0x3ffd11f0  
A2      : 0x3ffb55f8  A3      : 0x3ffb22b0  A4      : 0x00000000  A5      : 0x3ffb2300  
A6      : 0x00000001  A7      : 0x00000000  A8      : 0x8008b41d  A9      : 0x3ffd1230  
A10     : 0x00000000  A11     : 0x0002000c  A12     : 0x00000000  A13     : 0x3ffb6484  
A14     : 0x00000001  A15     : 0x00000000  SAR     : 0x00000010  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000000  LBEG    : 0x4000c2e0  LEND    : 0x4000c2f6  LCOUNT  : 0x00000000  

Backtrace: 0x4008124f:0x3ffd11f0 0x4008135f:0x3ffd1230 0x40080ff1:0x3ffd1260 0x400817ee:0x3ffd1290 0x4008180d:0x3ffd12c0 0x40081289:0x3ffd12f0 0x400841f1:0x3ffd1330 0x40081845:0x3ffd1360 0x40081386:0x3ffd1390 0x40081386:0x3ffd13c0 0x40082c36:0x3ffd13f0 0x40082d5a:0x3ffd1420 0x40082721:0x3ffd1450 0x40082721:0x3ffd1480 0x400841f1:0x3ffd14b0 0x40081ba1:0x3ffd14e0 0x40081386:0x3ffd1510 0x40082721:0x3ffd1540 0x400828b2:0x3ffd1570 0x40082a16:0x3ffd15a0 0x400828de:0x3ffd15d0 0x400866c5:0x3ffd1600 0x40081386:0x3ffd1630 0x40082a16:0x3ffd1660 0x40081651:0x3ffd1690 0x400828b2:0x3ffd16c0 0x40082a49:0x3ffd16f0 0x40081386:0x3ffd1720 0x400870f2:0x3ffd1750 0x40087115:0x3ffd1790 0x40081386:0x3ffd17c0 0x40081386:0x3ffd17f0 0x40081242:0x3ffd1820 0x4008121a:0x3ffd1850 0x40080f85:0x3ffd1890 0x400810be:0x3ffd18d0 0x40081309:0x3ffd1910 0x40081309:0x3ffd1940 0x4008103e:0x3ffd1970 0x40081309:0x3ffd19a0 0x40080fa3:0x3ffd19d0 0x400810be:0x3ffd1a10 0x40081309:0x3ffd1a50 0x40081309:0x3ffd1a80 0x400846a6:0x3ffd1ab0 0x40080fa3:0x3ffd1ae0 0x400810be:0x3ffd1b20 0x400846a6:0x3ffd1b60 0x40080ff1:0x3ffd1b90 0x40083785:0x3ffd1bc0 0x40081309:0x3ffd1bf0
```

I'm trying to decode it with the toolchain decoder, but no clue yet why it crashes when using those more advanced data structures from Assemblyscript.

About the `--importMemory`, was just me trying random stuff to get it working haha Btw, hanks for support folks, I'm trying all proposed things here.

**Comment by vshymanskyy** [comment]
I suggest you to print the whole buffer in hex (not until the `\0`. Then will see if all the data is correct, and its encoding.
Then you can try to adapt some simplified algorithms to convert the encoding to UTF8 or ASCII string.

**Comment by alvarowolfx** [comment]
Just to keep you on the loop, I tried the same thing with Rust and it worked fine to pass pointers and strings. Still trying to figure out why String.UT8 crashes the ESP32 when used on the Assemblyscript code.

What I used for Serial.log and Rust

Rust:
```
#[link_name = "serialLog"]      fn unsafe_println(out: *const u8);

pub fn serial_log     (out: &str)       { 
  unsafe {     
    unsafe_println(out.as_bytes().as_ptr() as *const u8)
  }
}
```

C/Arduino:
```
m3ApiRawFunction(m3_arduino_log)
{
  m3ApiGetArgMem(const char *, out)
      Serial.println(out);
  m3ApiSuccess();
}
m3_LinkRawFunction(module, arduino, "serialLog", "v(*)", &m3_arduino_log);
```

**Comment by MaxGraey** [comment]
Plz share your AssemblyScript as well

**Comment by alvarowolfx** [comment]
I think I tried all combinations of passing just string, string with length, pointer to string, String.UTF8 methods, etc

Here is my Assemblyscript 
```
@external("arduino", "millis")
declare function millis(): u32;

@external("arduino", "delay")
declare function delay(ms: u32): void;

@external("arduino", "pinMode")
declare function pinMode(pin: u32, mode: u32): void;

@external("arduino", "digitalWrite")
declare function digitalWrite(pin: u32, value: u32): void;

@external("arduino", "getPinLED")
declare function getPinLED(): u32;

@external("arduino", "serialLog")
declare function println(output: string): void;
//declare function println(output: usize, len: i32): void;
//declare function println(output: string, len: i32): void;

function log(out: string): void {
  //println(out, out.length)  // <- Only first letter arrives, but lenght is correct
  println(out); // <- Only first letter arrives
  String.UTF8.encode(out, true) // <- Just referencing this crashes
  //println(changetype<usize>(String.UTF8.encode(out, true)), String.UTF8.byteLength(out, true)); // <- Crash
  //println(String.UTF8.encode(out, true), String.UTF8.byteLength(out, true)); // <- Crash
}

const LOW: u32 = 0;
const HIGH: u32 = 1;

const INPUT: u32 = 0x0;
const OUTPUT: u32 = 0x1;
const INPUT_PULLUP: u32 = 0x2;

let LED: u32 = -1;

function setup(): void {
  LED = getPinLED();
  pinMode(LED, OUTPUT);
}

function run(): void {
  log('From Web Assembly')
  digitalWrite(LED, HIGH);
  delay(1000);
  digitalWrite(LED, LOW);
  delay(1000);
}

/*
 * Entry point
 */
export function _start(): void {
  setup();
  while (1) run();
}
```

**Comment by MaxGraey** [comment]
It should be:
```ts
@external("arduino", "serialLog")
declare function println(ptr: usize): void;

function log(str: string): void {
  println(changetype<usize>(String.UTF8.encode(str, true)));
}
```
Also try to build with `--runtime stub` instead `--runtime half` flag.

**Comment by alvarowolfx** [comment]
I tried that combination again here and still crashes @MaxGraey 😢 

Like I mentioned before, just using String.UTF8 or any ArrayBuffer/Uint8Array types crashes it, is really weird.

**Comment by MaxGraey** [comment]
Did you try with `--runtime stub` flag? I guess it could relate to memory allocation. `stub` uses simple bump allocator.

**Comment by alvarowolfx** [comment]
Sorry, forgot to mention that, I tried with `--runtime stub` too and made some more tests with it, but even without the string passing, WASM3 crashes with `stub` runtime. 

So my guess that this is somehow related to the memory allocation modes of Assemblyscript not fitting well with WASM3. But I have no clue how to test that properly. I probably need to forget a bit of hardware and test WASM3 locally on my computer to see what is going on.  Or, just use Rust for now.

**Comment by MaxGraey** [comment]
Using `String.UTF8.encode` on AS add runtime and allocation in heap. For Rust it seems it just put in data segment because don't do anything complex. Did you try allocate something on Rust as well? Like `let v: Vec<i32> = Vec::with_capacity(16);`

**Comment by vshymanskyy** [comment]
I will try to get to this in a few hours

**Comment by alvarowolfx** [comment]
I'll try that later @MaxGraey. Cool @vshymanskyy , I pushed my repository if that helps. Hope that I'm not missing something stupid 😅 

https://github.com/alvarowolfx/wasm-arduino-wifi

**Comment by vshymanskyy** [comment]
I hope you're missing something stupid. Otherwise we may run into more serious problems )

**Comment by alvarowolfx** [comment]
Weel, worst-case scenario, seems like I can move forward using Rust. I'm going to build a demo using Web Assembly as a way for basic rules change and/or OTA mechanism 🤓 .

**Comment by vshymanskyy** [comment]
Yup. I was also thinking about this approach.
Also, please note that the Arduino API as it's implemented here - is just an example. I really do not support this way of programming, and I think we have to come up with a better API (i.e. EventDriven, resource-oriented)

**Comment by vshymanskyy** [comment]
@alvarowolfx it looks like @MaxGraey suggestions worked for me!

With:
```ts
@external("arduino", "log")
declare function println(ptr: usize, len: i32): void;

function log(str: string): void {
  println(changetype<usize>(String.UTF8.encode(str)), String.UTF8.byteLength(str));
}
```
compiled with `--runtime stub` and:
```cpp
m3ApiRawFunction(m3_arduino_log)
{
  m3ApiGetArgMem(const byte *, out)
  m3ApiGetArg(int32_t, out_len)

  printf("== buff %p\n", out);

  byte buff[out_len+1];
  memcpy(buff, out, out_len);
  buff[out_len] = '\0';

  Serial.println((char*)buff);
  m3ApiSuccess();
}
```
I get:
![image](https://user-images.githubusercontent.com/1506708/72761585-1db25f00-3be5-11ea-8b5d-2f2f964f9e2d.png)

**BUT.** If you look at the string address, you can see that it constantly grows. As @MaxGraey mentioned, stub uses a dummy allocator (not able to `free` memory). So it will `Trap` when it reaches `WASM_MEMORY_LIMIT`. Just after a few minutes in my case.

**Comment by vshymanskyy** [comment]
Next, if I do:
```cpp
m3ApiRawFunction(m3_arduino_log)
{
  m3ApiGetArgMem(const byte *, out)
  m3ApiGetArg(int32_t, out_len)

  out_len *= 2; // Each symbol is 2 bytes

  byte buff[out_len+1];
  int len = sizeof(buff)-1;
  UTF16BEToUTF8(buff, &len, out, &out_len);
  buff[len] = '\0';

  Serial.println((char*)buff);

  m3ApiSuccess();
}
```
with
```ts
@external("arduino", "log")
declare function println(ptr: usize, len: i32): void;

function log(str: string): void {
  println(changetype<usize>(str), str.length)
}
```
It also works. This time, the conversion is performed on Host side. (I omitted the `UTF16BEToUTF8` function).

This time, the pointer is not growing because we're just printing constant strings. But it will grow as soon as you do something like this in the loop:
```ts
  const t: u32 = millis();
  log('Time: ' + t.toString());
```

So we have a solution for printing constant strings only.
As soon as we allocate dynamic memory (and with `--runtime half`, I mean) `AssemblyScript` (and other languages) starts relying on 64-KiB memory pages. So you need to remove the `WASM_MEMORY_LIMIT` restriction.

After removing the limit and running the app (`--runtime half`, with dynamic allocation), I got a crash.
I tried increasing `NATIVE_STACK_SIZE` to 128KiB and now it just hangs.
This may indicate that either there's some bug in `Wasm3`, or we need to develop a custom Linear Memory allocator for embedded systems.

I hope this gives at least some answers.

**Comment by vshymanskyy** [comment]
@alvarowolfx I can also suggest you to run such tests on x86 first, it's much easier to debug. For example, I just found the we have some assertions failing in this scenario when running in Debug mode. Setting up Debugging on ESP32 is a bit more tricky ;)

**Comment by vshymanskyy** [comment]
I think we're also hitting this bug here: https://github.com/wasm3/wasm3/issues/40

**Comment by vshymanskyy** [comment]
There's hope ;) With some workarounds for wasm3/wasm3#40 I got it working:
![image](https://user-images.githubusercontent.com/1506708/72767256-324e2180-3bfc-11ea-9a1a-0f63c07e30b1.png)

**Comment by vshymanskyy** [comment]
@alvarowolfx I have a small present for you:
[AssemblyScript_Log.zip](https://github.com/wasm3/wasm3-arduino/files/4091525/AssemblyScript_Log.zip)
:smile:

**Comment by alvarowolfx** [comment]
Wow, awesome guys, I'll test that later today. I noticed that you guys are using a Nightly version of Assemblyscript, so maybe that fixes some of my issues too. But really excited with what you guys showed here.

**Comment by MaxGraey** [comment]
But so simplicity and code compactness I recommend expose different print functions like:
```ts
// declare function printStringUTF8(ptr: usize, len: i32): void; not useful for AS
declare function printStringUTF16(ptr: usize, len: i32): void;
declare function printInt32UTF16(value: i32, unsigned: bool): void;
declare function printInt64UTF16(value: i64, unsigned: bool): void;
```
and do all conversion stuff on host side

**Comment by vshymanskyy** [comment]
I was also able to run it on RTL8711.
Basically it should run on any device that can afford to allocate 1-2 pages of Wasm Linear Memory (i.e. have `>= 128KiB` RAM)

**Comment by alvarowolfx** [comment]
woohoo, got basic Wifi connection and status working. 

<img width="1792" alt="Screen Shot 2020-01-21 at 6 18 38 PM" src="https://user-images.githubusercontent.com/1615543/72848725-98c65480-3c7b-11ea-8ac2-187c8649b0c5.png">

So now passing strings are working just fine. Thanks guys.

Now I'm trying to figure out how to pass strings back from the native code. I was starting trying to return the string with the device's local IP back to AS. Any hints on how to do that? 

I was hitting some memory access errors ( `[trap] out of bounds memory access`) when trying to return a pointer from native code and converting that pointer to an Arraybuffer or something else on AS side :


C/Native Code
```
m3ApiRawFunction(m3_arduino_wifi_local_ip)
{
  m3ApiReturnType(const byte *);

  const char *ip = WiFi.localIP().toString().c_str();
  uint32_t len = strlen(ip);
  byte local_ip[len + 1];
  memcpy(local_ip, ip, len);
  local_ip[len] = '\0';
  Serial.println(ip);

  m3ApiReturn(local_ip);
}

m3_LinkRawFunction(module, wifi, "wifiLocalIp", "*()", &m3_arduino_wifi_local_ip);
```

AS:
```
@external("wifi", "wifiLocalIp")
declare function wifiLocalIp(): usize;

function localIp(): string {
  const ptr = wifiLocalIp()
  log(ptr.toString()) // <- This seems to print the memory address

 /* All above here all fail with memory access error */ 
  //const arr = new ArrayBuffer(10) 
  //memory.copy(changetype<usize>(arr), ptr, 10)
  //const arr = load<ArrayBuffer>(ptr)
  // const dt = String.UTF8.(ip, 10, true)
  //const ip = load<string>(ptr)
  //return dt
}
```

**Comment by MaxGraey** [comment]
Pass string from host to wasm will require little bit more efforts. First of all you should change from `--runtime half` to `--runtime full`. This will export `__alloc` and `__retain` which should used [something like this](https://github.com/AssemblyScript/assemblyscript/blob/master/lib/loader/index.js#L120). And after in pseudocode:
```ts
fun allocString(str: wstring): u32 {
  // call wasm.__alloc here
}

fun wifi_local_ip(): u32 {
   var ptr = wasm.__retain(allocString("some string"));
   return ptr;
}
```

**Comment by vshymanskyy** [comment]
@MaxGraey isn't there a way to construct a string from UTF8 buffer?

**Comment by vshymanskyy** [comment]
Also, VM could allocate a buffer and pass it to the Native code to fill-in.

**Comment by vshymanskyy** [comment]
@alvarowolfx  this worked for me:
```cpp
m3ApiRawFunction(m3_arduino_getString)
{
    m3ApiGetArgMem(char *, out)
    m3ApiGetArg(int32_t, out_len)

    static const char buff[] = "Тест юнікоду.";
    memcpy(out, buff, min(sizeof(buff), out_len));

    m3ApiSuccess();
}
```

```ts
export function getString(): string {
  const arr = new ArrayBuffer(64)
  _getString(changetype<usize>(arr), 64)
  return String.UTF8.decode(arr, true)
}
```
Here is the full example (note that it has separate `app.ts` and `arduino.ts` now): 
[AssemblyScript_Log_2.zip](https://github.com/wasm3/wasm3-arduino/files/4094593/AssemblyScript_Log_2.zip)

P.S. @MaxGraey is suggesting a more general solution I think.. need to investigate.

**Comment by alvarowolfx** [comment]
Awesome folks, everything working now. I feel like now I can do anything from now on haha 💯 

Thank you so much for your support. I'll play some more with that and share my demo with you when I have something more interesting.
